### PR TITLE
Keystone business logic implementation & ui elaboration

### PR DESCRIPTION
**Pull request by Milan-Cerovsky**
> **Note**
>This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [ ] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [ ] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [ ] Update **documentation** as appropriate (e.g [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), [**CHANGELOG.md**](../blob/main/CHANGELOG.md), etc.)
- [ ] **Run the app** and try the changes
- [ ] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

> **Note**
> It is good practice to provide before and after UI  **screenshots** in the description of this PR. This is only applicable for changes that modify the UI.

# Reviewer

- [ ] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [ ] Perform an **ad hoc review**[^5]
- [ ] Review the **automated tests**
- [ ] Review the **manual tests**
- [ ] Review the **documentation**, [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), etc. as appropriate
- [ ] **Run the app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -17,7 +17,7 @@ android.useAndroidX=true
 org.gradle.kotlin.dsl.skipMetadataVersionCheck=false
 
 # Kotlin compiler warnings can be considered errors, failing the build.
-ZCASH_IS_TREAT_WARNINGS_AS_ERRORS=true
+ZCASH_IS_TREAT_WARNINGS_AS_ERRORS=false
```
Comment (HonzaR) [inline gradle.properties:20]: Revert this change, please.

File: gradle.properties [hunk 2]
```diff
@@ -138,7 +138,7 @@ ZCASH_EMULATOR_WTF_API_KEY=
 # Optional absolute path to a Zcash SDK checkout.
 # When blank, it pulls the SDK from Maven.
 # When set, it uses the path for a Gradle included build.  Path can either be absolute or relative to the root of this app's Gradle project.
-SDK_INCLUDED_BUILD_PATH=
+SDK_INCLUDED_BUILD_PATH=/Users/milancerovsky/Developer/zcash/zcash-android-wallet-sdk
```
Comment (HonzaR) [inline gradle.properties:141]: Revert this change, please.

File: ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/ZashiCheckbox.kt [hunk 3]
```diff
@@ -104,6 +77,39 @@ fun ZashiCheckbox(
...
+            scaleOut(
+                spring(
+                    stiffness = Spring.StiffnessHigh,
+                    dampingRatio = Spring.DampingRatioMediumBouncy
+                )
+            )
+        ) {
+            Image(
+                painter = painterResource(R.drawable.ic_zashi_checkbox_checked),
+                contentDescription = ""
```
Comment (HonzaR) [inline ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/ZashiCheckbox.kt:107]: Ideally, null instead of ""

File: ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/ZashiCheckbox.kt [hunk 4]
```diff
@@ -104,6 +77,39 @@ fun ZashiCheckbox(
     }
 }
 
+@Composable
+fun ZashiCheckboxIndicator(isChecked: Boolean) {
+    Box {
+        Image(
+            painter = painterResource(R.drawable.ic_zashi_checkbox),
+            contentDescription = ""
```
Comment (HonzaR) [inline ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/ZashiCheckbox.kt:85]: Ideally, null instead of ""

File: ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/ZashiQr.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        shape = RoundedCornerShape(ZashiDimensions.Radius.radius4xl),
+        border = BorderStroke(width = 1.dp, color = ZashiColors.Surfaces.strokePrimary),
+        color = ZashiColors.Surfaces.bgPrimary orDark ZashiColors.Surfaces.bgAlt
+    ) {
+        Box(
+            modifier = Modifier.padding(all = 6.dp)
+        ) {
+            Image(
+                bitmap = bitmap,
+                contentDescription = "",
```
Comment (HonzaR) [inline ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/ZashiQr.kt:42]: Ideally, null instead of ""

File: ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/listitem/ZashiContactListItem.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                fontWeight = FontWeight.SemiBold,
+            )
+            if (state.isShielded) {
+                Image(
+                    modifier =
+                    Modifier
+                        .align(Alignment.BottomEnd)
+                        .size(24.dp),
+                    painter = painterResource(id = R.drawable.ic_address_book_shielded),
+                    contentDescription = ""
```
Comment (HonzaR) [inline ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/listitem/ZashiContactListItem.kt:100]: Ideally, null instead of ""

**Review by HonzaR — COMMENTED** [review]
I took a first look at these changes and left a few comments.

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/datasource/ProposalDataSource.kt [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+        shieldingThreshold: Zatoshi,
+        memo: String,
+        transparentReceiver: String?
+    ): ShieldTransactionProposal =
+        withContext(Dispatchers.IO) {
+            getOrThrow {
+                val newProposal =
+                    synchronizerProvider.getSynchronizer().proposeShielding(
+                        account = account.sdkAccount,
+                        shieldingThreshold = Zatoshi(DEFAULT_SHIELDING_THRESHOLD),
```
Comment (Milan-Cerovsky) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/common/datasource/ProposalDataSource.kt:125]: self review: not propagating method arguments

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/usecase/CreateKeystoneProposalUseCase.kt [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    private val keystoneProposalRepository: KeystoneProposalRepository,
+    private val navigationRouter: NavigationRouter
+) {
+    suspend operator fun invoke(zecSend: ZecSend) {
+        try {
+            keystoneProposalRepository.createProposal(zecSend)
+            keystoneProposalRepository.createPCZTFromProposal()
+            navigationRouter.forward(ReviewKeystoneTransaction)
+        } catch (e: Exception) {
+            keystoneProposalRepository.clear()
```
Comment (Milan-Cerovsky) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/common/usecase/CreateKeystoneProposalUseCase.kt:18]: self review: no need for catching exception here

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/usecase/DeriveKeystoneAccountUnifiedAddressUseCase.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+import cash.z.ecc.android.sdk.tool.DerivationTool
+import co.electriccoin.zcash.ui.common.provider.PersistableWalletProvider
+import co.electriccoin.zcash.ui.screen.addressbook.viewmodel.ADDRESS_MAX_LENGTH
+import com.keystone.module.ZcashAccount
+
+class DeriveKeystoneAccountUnifiedAddressUseCase(
+    private val persistableWalletProvider: PersistableWalletProvider
+) {
+    suspend operator fun invoke(account: ZcashAccount): String {
```
Comment (Milan-Cerovsky) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/common/usecase/DeriveKeystoneAccountUnifiedAddressUseCase.kt:11]: self review: default dispatcher

File: ui-lib/src/main/java/co/electriccoin/zcash/global/WalletCoordinator.kt [hunk 10]
```diff
@@ -17,6 +17,8 @@ fun WalletCoordinator.Companion.newInstance(
         persistableWallet =
             flow {
                 emitAll(persistableWalletPreference.observe(encryptedPreferenceProvider()))
-            }
+            },
+        accountName = "",
```
Comment (HonzaR) [inline ui-lib/src/main/java/co/electriccoin/zcash/global/WalletCoordinator.kt:21]: Let's move `accountName` and `keySource` into a globally accessible constants

**Review by HonzaR — APPROVED** [review]
Tested only. With few comments inline. Will do another posthoc review.

**Comment by HonzaR** [comment]
Merging this for passing the Google review process as soon as possible. Filed a follow-up for the failing tests: #1713
